DataViews: Add comparison operators for filtering

chihsuan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of WOOA7S-123


## Proposed Changes

Add the following comparison operators for filtering:

- Less than
- Greater than
- Less than or equal
- Greater than or equal

Changes:

- Introduced new operators in constants and types.
- Implemented filtering logic in filter-and-sort-data-view.
- Updated filter component to handle new operators.


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Some use cases in Woo Analytics require filters that accept user-provided input — like "Price greater than 100",  or "Order Total less than 100". 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `yarn dataviews:storybook:start`
* Go to http://localhost:57863/?path=/story/dataviews-dataviews--default
* Select the `Satellites` filter
* Select operator `greater than`
* Enter `10`
* Confirm that the filter applied correctly
* Try other operators, e.g. `less than`


https://github.com/user-attachments/assets/5f07feca-fc5c-481b-bb03-da642a955616



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
Add comparison operators for filtering: less than, greater than, less than or equal, and greater than or equal

- Introduced new operators in constants and types.
- Implemented filtering logic in filter-and-sort-data-view.
- Updated filter component to handle new operators.
- Enhanced tests to cover new filtering functionality for integers and datetimes.

louwie17 commented on packages/dataviews/src/field-types/integer.tsx:
```diff
@@ -8,7 +8,14 @@ import type {
 } from '../types';
 import { renderFromElements } from '../utils';
-import { OPERATOR_IS, OPERATOR_IS_NOT } from '../constants';
+import {
+	OPERATOR_IS,
+	OPERATOR_IS_NOT,
+	OPERATOR_LESS_THAN,
+	OPERATOR_GREATER_THAN,
+	OPERATOR_LESS_THAN_OR_EQUAL,
+	OPERATOR_GREATER_THAN_OR_EQUAL,
+} from '../constants';
```
You could import the `SINGLE_SELECTION_OPERATORS` here instead right?

chihsuan replied:
Yes, but we will add more operators in the future, so I think it's better to import what is actually used and not to use SINGLE_SELECTION_OPERATORS now. Does that make sense?

louwie17 commented on packages/dataviews/src/field-types/integer.tsx:
```diff
@@ -34,7 +41,14 @@ function isValid( value: any, context?: ValidationContext ) {
 }
 
-const operators: Operator[] = [ OPERATOR_IS, OPERATOR_IS_NOT ];
+const operators: Operator[] = [
+	OPERATOR_IS,
+	OPERATOR_IS_NOT,
+	OPERATOR_LESS_THAN,
+	OPERATOR_GREATER_THAN,
+	OPERATOR_LESS_THAN_OR_EQUAL,
+	OPERATOR_GREATER_THAN_OR_EQUAL,
+];
```
Should we also be adding them here: https://github.com/Automattic/wp-calypso/blob/0c7ece89853d6620c4855bde937fa2e752c840a2/packages/dataviews/src/field-types/datetime.tsx#L31 ?

chihsuan replied:
Good question. I'm exploring what the best way to do this is.

For the datetime field, we need to add these operators (Figma: hdK9XLiOefugy2bxalKBN1-fi-2806_89215):

<img width="331" alt="Screenshot 2025-05-20 at 09 21 44" src="https://github.com/user-attachments/assets/32c07a97-a106-44da-8d0b-51a265616b3c" />

The `After (inc)` and `Before (inc)` operators are the same as `Greater than or equal to` and `Less than or equal to` operators. However, the design requires us to show them in a different title in the UI,  which does feel better from a user experience perspective.
 
We could potentially create a mapping of operators to their corresponding titles. I'd appreciate any suggestions you might have.

chihsuan replied:
I've explored the approach. I think there are two solutions:

1. Add new operator keys to the Operator enum and constants. And then update the filtering logic to treat these new operators as aliases for existing ones (e.g., `afterInc` acts like `greaterThanOrEqual`).

Pros
- Operator label is globally consistent: anywhere the new operator is used, the label is the same.
- UI logic remains simple: always use the global OPERATORS map for labels.
- Explicit mapping: the intent is clear in the codebase.

Cons
- Requires changes to core types, constants, and filtering logic for every new operator.
- Potential for confusion: multiple operator keys with identical logic but different names.

2. Add a new prop `operatorLabels` to field type (e.g., datetime)

```js
export default {
  // ...other field type config,
  type: 'datetime',
  // ...,
  operatorLabels: {
    greaterThanOrEqual: __('After (inc)'),
    lessThanOrEqual: __('Before (inc)'),
  },
};
```

Pros
- No changes to core types or filtering logic.
- No risk of logic duplication—all filtering continues to use the same operator keys.
- Adding new field types is fully decoupled: only the field type file and (once) the UI label logic need to be touched.
- Minimal code changes: only the field type and the UI rendering logic.
- No risk of breaking existing code or introducing bugs in filtering.

Cons
- Slightly more complex UI logic: the UI must check for a field-specific label.
- Operator label customization is only possible via the field type, not globally


Since we only want to change the label for a specific field type (e.g., datetime), and the logic is identical, I think the second approach is better. Consumers can also customize the operator labels for their own use cases.

What do you think?

chihsuan replied:
I also noticed that in Woo Analytics, we want to display "Equal to" for the integer field type for the "is" operator. 

<img width="511" alt="Screenshot 2025-05-20 at 13 59 10" src="https://github.com/user-attachments/assets/679495ad-30e2-47a7-ba22-11e483b49238" />

retrofox replied:
Is this something to consider addressing in a separate task?

chihsuan replied:
Sounds good! : )  I just wanted to gather some feedback on this.

louwie17 reviewed (commented):
This overall looks good to me, I just left a couple small inline comments around using the operators for the date time field type, and an import related thing.
Happy to approve this, once this is based off of `trunk`.

chihsuan committed:
Update dataviews-filters styles to use min-width for popover content

chihsuan commented on packages/dataviews/src/components/dataviews-filters/style.scss:
```diff
@@ -22,7 +22,7 @@
 
 	.components-popover__content {
-		width: 230px;
+		min-width: 230px;
```
@Automattic/dotcom-design The filter popover container is set to 230px fixed width. With the new operators, the filter operator text is getting longer. When the filter operator text is long, the scrollbar appears, which can create a poor user experience. 

Therefore, I changed the css width to `min-width` to allow the filter popover to expand to the width of the operator text. 


Before:
<img width="391" alt="Screenshot 2025-05-19 at 17 59 30" src="https://github.com/user-attachments/assets/6e7981a0-8f9c-4869-a191-a3d1629d933a" />

After:

<img width="391" alt="Screenshot 2025-05-20 at 09 49 14" src="https://github.com/user-attachments/assets/eeeb855c-e49a-43bc-bd79-945e718cfe48" />

However, this might make the filter excessively wide if the field name is very long. I'm considering whether we should implement a max-width for the name, but I'm unsure if this is the optimal approach. Additionally, I'm uncertain about what value to set for the max-width. Could you provide some guidance on this?  🙏 

<img width="373" alt="Screenshot 2025-05-20 at 09 58 20" src="https://github.com/user-attachments/assets/b7498e5c-d14a-4579-9581-0853427223bf" />

jasmussen replied:
That's an interesting challenge, thank you for surfacing it, and using the broader ping like you did.

The challenge is interesting not only for the popover, but for the filter chip itself, which could get really long. How does it wrap? Should it be elided? Notably mobile comes to mind. 

One quick instinct is: we should keep a fixed-width, modals that grow in size can in some contexts create layout shifts. But we should allow the title field to wrap, so it can grow vertically. We should also not elide in the modal, but we should consider setting a maxwidth on the filter chips (if there isn't already). Not urgent, so feel free to open an issue about it [here](https://linear.app/a8c/team/DS/triage), and use the "General Design System issue" template. but in the shown example we could have something like "Satellites is greater than or ...". You can always tap it to see the full title in the popover, is why that could work.

chihsuan replied:
Thanks for the feedback. @jasmussen  I'll open an issue about it. 🙌 

> in the shown example we could have something like "Satellites is greater than or ...". You can always tap it to see the full title in the popover, is why that could work.

I'll make the changes to the PR.

chihsuan replied:
Created issue: [DS-217](https://linear.app/a8c/issue/DS-217/design-system-dataviews-define-behavior-for-long-text-in-filter).

Also, applied a temporary fix.

<img width="350" alt="Screenshot 2025-05-20 at 17 26 12" src="https://github.com/user-attachments/assets/ae169308-0ba8-4919-b681-0a84b43c5d6b" />

jasmussen replied:
Curious if others have thoughts, but I think this can work.

retrofox replied:
Does it make sense to add a `title` property there?

<img width="360" alt="image" src="https://github.com/user-attachments/assets/4df6798c-6ca2-489d-99a4-0933a20efc45" />

chihsuan committed:
Fix types

chihsuan committed:
Update REDAME.md

chihsuan committed:
Add changelog

chihsuan committed:
Update filter name and select style

- Updated the OperatorSelector component to include a new class for the select control.
- Adjusted the width of the popover content and added styles for the summary operators filter select to ensure proper display and prevent shrinking.

retrofox commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -152,6 +156,58 @@ const FilterText = ( {
+	if ( filterInView?.operator === OPERATOR_GREATER_THAN ) {
+		return createInterpolateElement(
+			sprintf(
+				/* translators: 1: Filter name. 3: Filter value. e.g.: "Author is not: Admin". */
```
Can you update the translator's hint to something more in line with `is greater than`?

chihsuan replied:
Yes, definitely. 😓 All fixed in 72a6c8eb9dad46b385fa529282a665f9ef03b189.

retrofox commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -152,6 +156,58 @@ const FilterText = ( {
+	if ( filterInView?.operator === OPERATOR_LESS_THAN_OR_EQUAL ) {
+		return createInterpolateElement(
+			sprintf(
+				/* translators: 1: Filter name. 3: Filter value. e.g.: "Author is less than or equal to: Admin". */
```
Similar to this: I'm uncertain if `Author is less than or equal to: Admin` makes sense. A phrase like `Price is less than or equal ...` probably fits better.

retrofox commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -152,6 +156,58 @@ const FilterText = ( {
+	if ( filterInView?.operator === OPERATOR_GREATER_THAN_OR_EQUAL ) {
+		return createInterpolateElement(
+			sprintf(
+				/* translators: 1: Filter name. 3: Filter value. e.g.: "Author is greater than or equal to: Admin". */
```
similar :-)

retrofox commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -152,6 +156,58 @@ const FilterText = ( {
+	if ( filterInView?.operator === OPERATOR_LESS_THAN ) {
+		return createInterpolateElement(
+			sprintf(
+				/* translators: 1: Filter name. 3: Filter value. e.g.: "Author is not: Admin". */
```
similar to the comment below

retrofox commented on packages/dataviews/src/components/dataviews-filters/filter.tsx:
```diff
@@ -152,6 +156,58 @@ const FilterText = ( {
 		);
 	}
 
+	if ( filterInView?.operator === OPERATOR_LESS_THAN ) {
```
I wonder why we don't use a `switch ( filterInView?.operator ) {` statement. Just a comment, I think we should change that

chihsuan replied:
I'm fine with either way. If everyone is fine with this, I can change it. 🙂

retrofox replied:
I'd say not for the moment.

retrofox commented on client/sites/hooks/use-initialize-dataviews-page.ts:
```diff
@@ -1,6 +1,6 @@
 import { usePrevious } from '@wordpress/compose';
 import { useEffect, useRef } from 'react';
-import type { View } from '@wordpress/dataviews';
+import type { View } from '@automattic/dataviews';
```
I guess this change is fine. I'd like to know whether we want to change this, considering we may need to go back upstream in the future. cc @oandregal

chihsuan replied:
I changed this because lint/CI issues were reported. Dotcom Site is using `@automattic/dataviews`. Using `@wordpress/dataviews` will cause incompatible types.

https://github.com/Automattic/wp-calypso/blob/10f282589208f4270d6a93fbd9c9a59fc443d2b2/client/sites/components/sites-dataviews/index.tsx#L2

chihsuan committed:
Fix translators comments

retrofox reviewed (approved):
LGTM. Tested and works as expected, following the same approach used in other filters already implemented.

I'd suggest creating a follow-up based on https://github.com/Automattic/wp-calypso/pull/103525#discussion_r2096989610 to explore potential changes.

Let's wait in case others want to share additional feedback. Worst case, we can always keep iterating.
